drivers/java/src/main/java/com/rethinkdb/ast/ReqlAst.java

```diff
@@ -96,6 +170,47 @@ protected Object build() {
     }
 
     /**
+     * Runs this query via connection {@code conn} provided by connection pool,
+     * with options {@code runOpts} and returns an atom result
+     * or a sequence result as a cursor. The atom result representing a JSON object is converted
+     * to an object of type {@code Class<P>} specified with {@code pojoClass}. The cursor
+     * is a {@code com.rethinkdb.net.Cursor} which may be iterated to get a sequence of atom results
+     * of type {@code Class<P>}
+     * @param pool The connection pool to provide connection to run this query
+     * @param pojoClass The class of POJO to convert to
+     * @param <T> The type of result
+     * @param <P> The type of POJO to convert to
+     * @return The result of this query (either a {@code P or a Cursor<P>}
+     */
+    public <T, P> T run(ConnectionPool pool, Class<P> pojoClass) {
+    	  Connection conn = pool.leaseConnection();
+    	  T res = this.run(conn, pojoClass);
+    	  pool.returnConnection(conn);
+    	  return res;
```

Comment:
Hey there! Great work and contribution! This is pretty awesome! Just had a couple of questions here...

What happens when `this.run(conn, pojoClass)` throws an underlying exception? ReQL syntax `ReqlDriverError`? Seems like a connection could never be returned to the connection pool. Possible memory/connection leak?

Patterns like this (1. take connection, 2. use, 3. put back connection) puts on the onus on the caller of the pool to remember to put back a connection after usage (just my personal feeling of uneasiness here). I think you'll have to ultimately wrap `try{}catch{}finally{}` on each of these `run` statements to make this pattern work.

I have a few more comments but I'm currently on vacation right now but memory/connection leak is the most critical of my questions.

The second without much comment is the unboundedness of a "fixed connection pool" and weather or not all the orchestration (lease / use / return) is necessary when each connection is already thread-safe. Just use with a round-robin counter?

Sorry, I have to go!
Brian